pkg/cloudfabric-controller/controllerframework/controllerinstancemanager.go

```diff
@@ -125,29 +125,31 @@ func (cim *ControllerInstanceManager) addControllerInstance(obj interface{}) {
 
 	cim.mux.Lock()
 	klog.V(4).Infof("mux acquired addControllerInstance. CIM %v", cim.instanceId)
-	defer func() {
-		cim.mux.Unlock()
-		klog.V(4).Infof("mux released addControllerInstance. CIM %v", cim.instanceId)
-	}()
 
 	if cim.currentControllers == nil {
 		cim.currentControllers = make(map[string](map[string]v1.ControllerInstance))
 	}
 	existingInstancesForType, ok := cim.currentControllers[newControllerInstance.ControllerType]
+	var existingInstance v1.ControllerInstance
+	var ok1 bool
+
 	if !ok { // real new controller instance
 		cim.currentControllers[newControllerInstance.ControllerType] = make(map[string]v1.ControllerInstance)
 	} else { // check existing controller instance
-		existingInstance, ok1 := existingInstancesForType[newControllerInstance.Name]
-		if ok1 {
```

Comment:
You can leave the logic here and just add cim.mux.Unlock() before cim.updateControllerInstance

Reply:
In that case we might need to relock it and release the lock again to update controller instance mapping

Reply:
Notify handler and then does cim.updateControllerInstance again. If ok1 == true, notifyHandler will be called twice. Once in addControllerInstance, the other in updateControllerInstance